### ISSUE TITLE
Var-arg pattern

### ISSUE THREAD

**Issue by olafurpg** [scalacenter/scalafix#59]
Test from srewrite

```diff
 object Test {

   List(1, 2, 3, 4) match {
-    case List(1, 2, xs @ _*) =>
+    case List(1, 2, xs : _*) =>
       val ys: Seq[Int] = xs
       println(ys)
   }
-  val List(1, 2, x @ _*) = List(1, 2, 3, 4)
+  val List(1, 2, x : _*) = List(1, 2, 3, 4)

 }
```

**Comment by gabro** [comment]
What is this?

```scala
val List(1, 2, x: _*) = List(1, 2, 3, 4)
```

doesn't look like valid syntax, does it?

**Comment by olafurpg** [comment]
iirc, that will be the new syntax in dotty and scala 2.13 or 2.14

**Comment by gabro** [comment]
Oh ok, didn't know that, thanks 👍  Here it is in dotty https://github.com/lampepfl/dotty/blob/70a8426912f6f256536470a046cdf86827ab2c93/tests/pos/vararg-pattern.scala#L8

**Comment by gosubpl** [comment]
@olafurpg I will participate in the ScalaCenter Spree in Copenhagen. Can I try tackling this ticket? I would like to start early :)

**Comment by olafurpg** [comment]
Absolutely! @gosubpl Go ahead :) Don't hesitate to ask if you have any questions.

**Comment by gosubpl** [comment]
Is this a case of semantic rewrite (the one with `mirror`)? or just syntactic (I think semantic - please correct me if I am wrong :) ). Also, what is the purpose of the `.scalatex` file? Should I provide one?

**Comment by olafurpg** [comment]
This would be a syntactic rewrite, no mirror. Semantic rewrites require information from the compiler such as types, which is not necessary here.

> what is the purpose of the .scalatex file? 

That's for docs, there's no need for that in this rewrite.

**Comment by gosubpl** [comment]
Thanks for the clarification. Got it! I hope I will manage from here on my own :+1:

**Comment by gosubpl** [comment]
Hmmm... so this new syntax (with `:`) does not parse too well. Can I somehow hint the engine that I am using `Dotty` dialect for patch target?

**Comment by olafurpg** [comment]
Good question @gosubpl Can you try configuring `dialect = Dotty`? Let me know that works. It could be that the Patch api doesn't gracefully handle custom dialects, if so I can try to look into that before the Spree.

**Comment by gosubpl** [comment]
@olafurpg I will also keep digging, but for now it looks like the `dialect = Dotty` setting is being ignored.
```
Rewrite dialect: Dotty
SemanticRewriteSuite:88 [diffTest.unwrap(fixed)]: 
 object Test {

   List(1, 2, 3, 4) match {
     case List(1, 2, xs : _*) =>
       val ys: Seq[Int] = xs
       println(ys)
   }
   val List(1, 2, x : _*) = List(1, 2, 3, 4)

 }
```
and error
```
[info] - checkSyntax/VarArg.source: test *** FAILED *** (1 second, 508 milliseconds)
[info]   org.scalameta.invariants.InvariantFailedException: invariant failed:
[info] when verifying bounds.lo.nonEmpty.||(bounds.hi.nonEmpty)
[info] found that bounds.lo.nonEmpty is false
[info] and also bounds.hi.nonEmpty is false
[info] where bounds =
[info]   at org.scalameta.invariants.InvariantFailedException$.raise(Exceptions.scala:15)
```

if I do a rewrite that keeps the `@` sign instead of replacing with colon, everything works fine.

**Comment by gosubpl** [comment]
@olafurpg oooh, managed to change things a bit :)
After adding the `Dotty` dialect import in `SemanticRewriteSuite.parse`:
```
  private def parse(code: String): m.Tree = {
    import scala.meta._
    import scala.meta.dialects.Dotty
    code.parse[Source].get
  }
```

got different error:
```
[info] - checkSyntax/VarArg.source: test *** FAILED *** (303 milliseconds)
[info]   Fixed source code does not typecheck!
[info]   Message: Enclosing tree [9829] does not include tree [9824]
[info]   Reveal: package∙test∙{¶package∙test∙{¶∙object∙Test∙{¶¶∙∙∙List(1,∙2,∙3,∙4)∙match∙{¶∙∙∙∙∙case∙List(1,∙2,∙xs∙:∙_*)∙=>¶∙∙∙∙∙∙∙val∙ys:∙Seq[Int]∙=∙xs¶∙∙∙∙∙∙∙println(ys)¶∙∙∙}¶∙∙∙val∙List(1,∙2,∙x∙:∙_*)∙=∙List(1,∙2,∙3,∙4)¶¶∙}∙}¶∙}¶
[info]   Code: package test {
[info]   package test {
[info]    object Test {
[info]   
[info]      List(1, 2, 3, 4) match {
[info]        case List(1, 2, xs : _*) =>
[info]          val ys: Seq[Int] = xs
[info]          println(ys)
[info]      }
[info]      val List(1, 2, x : _*) = List(1, 2, 3, 4)
[info]   
[info]    } }
[info]    } (SemanticRewriteSuite.scala:228)
```

Is `Dotty` for sure OK with this code?

**Comment by olafurpg** [comment]
Seems like you are on great track @gosubpl . The `parse` method should use the `config.dialect(code).parse[Source]`

The tests verify that the "fixed" code type-checks with scalac. However, for this rewrite case it won't 
 even parse with scalac since the syntax is not supported by scalac.

For now, you can create a new test suite named `<RewriteName>Suite` that extends SyntacticRewriteSuite and use the `check(original, expected)` method. You may need to edit SyntacticRewriteSuite to support a custom dialect.

**Comment by gosubpl** [comment]
@olafurpg sure, but I am afraid that there is an issue with `scalameta` that it does not parse the `Dotty` dialect correctly. I am done for today, but will take `scalameta` for a ride tomorrow. After this is sorted out, I will move forward with the approach you suggested above.

**Comment by gosubpl** [comment]
OK, found the issue. `scalameta` correctly parses the expresion for `Dotty` dialect (at least since 2015 :) ) but before the parser is invoke, `scalafix` checks in line 91 of `SemanticRewriteSuite.scala` whether the code _type checks_ - https://github.com/scalacenter/scalafix/blob/master/scalafix-testkit/src/main/scala/scalafix/testkit/SemanticRewriteSuite.scala#L91

The typecheck is done using standard Scala compiler (see the `computeMirror` function in the same source). After commenting this line out:
```
    try {
//      typeChecks(diffTest.wrapped(fixed))  // <--- line 91 type check
      if (diffTest.checkSyntax) {
```
the problem goes away, as `Dotty` flavoured `scalameta` parses the code just fine.

Now question - @olafurpg is there a `computeMirror` counterpart for `Dotty` anywhere in `scalafix` ? How hard do you think would be to introduce one - I could try (in a separate ticket :) )

**Comment by olafurpg** [comment]
@gosubpl Thank you for tracking that down! I have actually been a bit unsatisfied with the test infrastructure lately, partly due to the errors you have encountered. The tests are pretty-much re-implementing what a build tool already provides. I finally took it on to refactor the entire test suite to compile the inputs and outputs with sbt, see https://github.com/scalacenter/scalafix/pull/165 This should make it easy to add a project that compiles with Dotty.

**Comment by gosubpl** [comment]
@olafurpg Lovely, many thanks. I will wait for https://github.com/scalacenter/scalafix/pull/165 to be merged and then resume working on this ticket :)

**Comment by gosubpl** [comment]
@olafurpg works like a charm. There is one issue only - the `VarArg` output does not compile with anything else than Dotty, so for now tests in project fail with:
```
[info] Compiling 1 Scala source to /home/janek/Documents/scala_spree/scalafix/scalafix-tests/input/target/scala-2.12/classes...
[error] [0] /scala/test/VarArg.scala:8: ')' expected but identifier found.
[error] [1] /scala/test/VarArg.scala:11: '=>' expected but '}' found.
[error] [2] /scala/test/VarArg.scala:13: ')' expected but identifier found.
[error] [3] /scala/test/VarArg.scala:14: '=' expected but '}' found.
[error] /scala/test/VarArg.scala: 8 [0], 11 [1], 13 [2], 14 [3]
```
How would you advise to work around that limitation?

**Comment by olafurpg** [comment]
Awesome @gosubpl ! We can add a new scalafix-tests/outputDotty project that compiles with Dotty. I will ping you on the PR when I add it.

**Comment by gosubpl** [comment]
Thanks. I gather that this will have `Dotty` configured as a compiler in `build.sbt` and the `.scalafix.conf` will identify `Dotty` as the dialect for `scala.meta`.

**Comment by olafurpg** [comment]
Correct, we can use sbt-dotty to compile an outputDotty with the dotty compiler. I don't think we will not need to fiddle with dialects, we'll see.

**Comment by olafurpg** [comment]
@gosubpl I opened #173 to add support for expected output that only compiles with Dotty. Instead of `scalafix-tests/output/src` use `scalafix-tests/output-dotty/src`

**Comment by gosubpl** [comment]
Many thanks @olafurpg !  Rebased & the doc error fixed (plus a minor sbt build issue for local builds)

**Comment by gosubpl** [comment]
This ticket can be closed now.

**Comment by olafurpg** [comment]
Closed by #179
